core/drivers/clk/clk.c

```diff
@@ -0,0 +1,280 @@
+TEE_Result clk_register(struct clk *clk)
+{
+	if (clk_check(clk))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	SLIST_INSERT_HEAD(&clk_list, clk, link);
+
+	clk_init_parent(clk);
+	clk_compute_rate(clk);
+
+	DMSG("Registered clock %s, freq %ld\n", clk->name, clk_get_rate(clk));
+
+	return TEE_SUCCESS;
+}
+
+unsigned long clk_get_rate(struct clk *clk)
+{
+	return clk->rate;
+}
+
+static TEE_Result clk_enable_no_count(struct clk *clk)
+{
+	if (clk->ops->enable) {
+		if (clk->ops->enable(clk))
+			return TEE_ERROR_GENERIC;
+	}
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result clk_enable(struct clk *clk)
+{
+	TEE_Result ret;
+	struct clk *parent;
+
+	if (clk->enabled_count == ULONG_MAX)
+		return TEE_ERROR_OVERFLOW;
+
+	if (clk->enabled_count) {
+		clk->enabled_count++;
+		return TEE_SUCCESS;
+	}
+
+	parent = clk_get_parent(clk);
+	if (parent)
+		clk_enable(parent);
+
+	ret = clk_enable_no_count(clk);
+	if (ret == TEE_SUCCESS)
+		clk->enabled_count = 1;
+
+	return ret;
+}
+
+static void clk_disable_no_count(struct clk *clk)
+{
+	if (clk->ops->disable)
+		clk->ops->disable(clk);
+}
+
+void clk_disable(struct clk *clk)
+{
+	struct clk *parent;
+
+	if (clk->enabled_count == 0) {
+		EMSG("Unbalanced clk_enable/disable");
+		return;
+	}
+
+	if (clk->enabled_count) {
+		clk->enabled_count--;
+		if (clk->enabled_count)
+			return;
+	}
+
+	clk_disable_no_count(clk);
+
+	parent = clk_get_parent(clk);
+	if (parent)
+		clk_disable(parent);
+
+	clk->enabled_count = 0;
+}
+
+int clk_is_enabled(struct clk *clk)
+{
+	if (clk->ops->is_enabled)
+		return clk->ops->is_enabled(clk);
+
+	return clk->enabled_count;
+}
+
+TEE_Result clk_set_rate(struct clk *clk, unsigned long rate)
+{
+	int was_enabled = 0;
+	unsigned long parent_rate = 0;
+
+	if (clk->flags & CLK_SET_RATE_GATE) {
+		was_enabled = clk_is_enabled(clk);
+		if (was_enabled)
+			clk_disable_no_count(clk);
+	}
+
+	/*
+	 * No need to check for clk->parent since this is checked at clock
+	 * registration
+	 */
+	if (clk->flags & CLK_SET_RATE_PARENT)
+		clk_set_rate(clk->parent, rate);
+
+	if (clk->ops->set_rate) {
+		if (clk->parent)
+			parent_rate = clk->parent->rate;
+
+		if (clk->ops->set_rate(clk, rate, parent_rate))
+			return TEE_ERROR_GENERIC;
+	}
+
+	if (clk->flags & CLK_SET_RATE_GATE && was_enabled)
+		clk_enable_no_count(clk);
+
+	clk_compute_rate(clk);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result clk_set_parent(struct clk *clk, unsigned int pidx)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	int was_enabled = 0;
+
+	if (pidx >= clk->num_parents || !clk->ops->set_parent)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	/* Requested parent is already the one set */
+	if (clk->parent == clk->parents[pidx])
+		return TEE_SUCCESS;
+
+	if (clk->flags & CLK_SET_PARENT_GATE) {
+		was_enabled = clk->enabled_count;
+		if (was_enabled)
+			clk_disable(clk);
+	}
+
+	if (clk->ops->set_parent(clk, pidx)) {
+		ret = TEE_ERROR_GENERIC;
+		goto err_out;
+	}
+
+	clk->parent = clk->parents[pidx];
+
+	if ((clk->flags & CLK_SET_PARENT_GATE) && was_enabled)
```

Comment:
testing `was_enabled` is sufficient.

Reply:
Fixed both cases in set_parent and set_rate